gitclone/checkout.go

```diff
@@ -0,0 +1,248 @@
+	InvalidCheckoutMethod CheckoutMethod = iota
+	// CheckoutNoneMethod only adds remote, resets repo, updates submodules
+	CheckoutNoneMethod
+	// CheckoutCommitMethod checks out a commit
+	CheckoutCommitMethod
+	// CheckoutTagMethod checks out a tag
+	CheckoutTagMethod
+	// CheckoutBranchMethod checks out a branch
+	CheckoutBranchMethod
+	// CheckoutPRMergeBranchMethod checks out a MR/PR in case a merge branch is available
+	CheckoutPRMergeBranchMethod
+	// CheckoutPRDiffFileMethod  checks out a MR/PR in case a diff file is available
+	CheckoutPRDiffFileMethod
+	// CheckoutPRManualMergeMethod checks out a MR
+	CheckoutPRManualMergeMethod
+	// CheckoutForkPRManualMergeMethod checks out a PR
+	CheckoutForkPRManualMergeMethod
+)
+
+// ParameterValidationError is returned when there is missing or malformatted parameter for a given parameter set
+type ParameterValidationError struct {
+	ErrorString string
+}
+
+// Error ...
+func (e ParameterValidationError) Error() string {
+	return e.ErrorString
+}
+
+// NewParameterValidationError return a new ValidationError
+func NewParameterValidationError(msg string) error {
+	return ParameterValidationError{ErrorString: msg}
+}
+
+// checkoutStrategy is the interface an actual checkout strategy implements
+type checkoutStrategy interface {
+	do(gitCmd git.Git, fetchOptions fetchOptions, fallback fallbackRetry) error
+}
+
+// X: required parameter
+// !: used to identify checkout strategy
+// _: optional parameter
+// ?: manual PR checkout method uses PRRepoURL to identify if it should be used,
+//    branchDest is a better candidate
+// |==================================================================================|
+// | params\strat| commit | tag | branch | manualMR | manualPR | autoMerge | autoDiff |
+// | commit      |  X  !  |     |        |  X       |          |           |          |
+// | tag         |        |  X !|        |          |          |           |          |
+// | branch      |  _     |  _  |  X !   |  X       |  X       |  X        |          |
+// | branchDest  |        |     |        |  X       |  X       |           |  X       |
+// | PRRepoURL   |        |     |        |  ?   !   |  X !     |    !      |    !     |
+// | PRID        |        |     |        |          |          |           |  X       |
+// | mergeBranch |        |     |        |          |          |  X !      |          |
+// |==================================================================================|
+
+func selectCheckoutMethod(cfg Config) CheckoutMethod {
+	isPR := cfg.PRRepositoryURL != "" || cfg.PRMergeBranch != "" || cfg.PRID != 0
+	if !isPR {
+		if cfg.Commit != "" {
+			return CheckoutCommitMethod
+		}
+
+		if cfg.Tag != "" {
+			return CheckoutTagMethod
+		}
+
+		if cfg.Branch != "" {
+			return CheckoutBranchMethod
+		}
+
+		return CheckoutNoneMethod
+	}
+
+	// ** PR **
+	isPrivateFork := isPrivate(cfg.PRRepositoryURL) && isFork(cfg.RepositoryURL, cfg.PRRepositoryURL)
```

Comment:
isFork() return value could be assigned to a var and it could be reused in line 99.